reaper.go

```diff
@@ -0,0 +1,185 @@
+//
+// Copyright (c) 2017 Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+package main
+
+import (
+	"fmt"
+	"os"
+	"os/exec"
+	"sync"
+
+	"github.com/opencontainers/runc/libcontainer"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/sys/unix"
+)
+
+type reaper struct {
+	sync.Mutex
+
+	chansLock     sync.RWMutex
+	exitCodeChans map[int]chan int
```

Comment:
In general, the invoking count of write lock is the same as the invoking count of reader lock.
they are both about the number of the registered pid multiply by 2.

the major lock of the reaper (reaper.Mutex) makes the creations of processes sequenced. it can be changed to RWMutex to allow them parallel.

Anyway, both are very trivial problems.

Reply:
@laijs I have updated the PR by using a read lock for new processes, so that they can be started in parallel. The reaper is a RW lock since we do want it to block until all read locks are done.
About the 2 different mutexes, this is the way to synchronize the reaper and processes in one hand, while in the other hand, we can ensure a map is not accessed at the same time from two different processes.